### ISSUE TITLE
CSS and HTML servers are breaking LSP

### ISSUE THREAD

**Issue by mniewrzal** [Microsoft/vscode#17289]
- VSCode Version:
Version 1.8.0
Commit 38746938a4ab94f2f57d9e1309c51fd6fb37553d
Date 2016-12-13T18:03:00.506Z
Shell 1.4.6
Renderer 53.0.2785.143
Node 6.5.0
- OS Version: Fedora 25

With latest version 1.8 vscode language servers CSS and HTML (and maybe others) are using completion response structure that is invalid from language server protocol definition point of view.
Example CompletionItem:
```json
        {
            "label": "ion-view ",
            "kind": 10,
            "documentation": "Child of ionNavView. A container for view content and any navigational and header bar information.",
            "range": {
                "start": {
                    "line": 2,
                    "character": 1
                },
                "end": {
                    "line": 2,
                    "character": 5
                }
            }
        }
```
There is no such field like `range` in CompletionItem [1]. Response should contain `textEdit` field or language server protocol definition should be updated.

[1] https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#completion-request

**Comment by dbaeumer** [comment]
We want to deprecate textEdit in favor of insertText and range since it was confusing if people provided a insertText and a textEdit. The new TS interface will look like this:

```typescript
	export class CompletionItem {

		/**
		 * The label of this completion item. By default
		 * this is also the text that is inserted when selecting
		 * this completion.
		 */
		label: string;

		/**
		 * The kind of this completion item. Based on the kind
		 * an icon is chosen by the editor.
		 */
		kind?: CompletionItemKind;

		/**
		 * A human-readable string with additional information
		 * about this item, like type or symbol information.
		 */
		detail?: string;

		/**
		 * A human-readable string that represents a doc-comment.
		 */
		documentation?: string;

		/**
		 * A string that should be used when comparing this item
		 * with other items. When `falsy` the [label](#CompletionItem.label)
		 * is used.
		 */
		sortText?: string;

		/**
		 * A string that should be used when filtering a set of
		 * completion items. When `falsy` the [label](#CompletionItem.label)
		 * is used.
		 */
		filterText?: string;

		/**
		 * A string or snippet that should be inserted in a document when selecting
		 * this completion. When `falsy` the [label](#CompletionItem.label)
		 * is used.
		 */
		insertText?: string | SnippetString;

		/**
		 * A range of text that should be replaced by this completion item.
		 *
		 * Defaults to a range from the start of the [current word](#TextDocument.getWordRangeAtPosition) to the
		 * current position.
		 *
		 * *Note:* The range must be a [single line](#Range.isSingleLine) and it must
		 * [contain](#Range.contains) the position at which completion has been [requested](#CompletionItemProvider.provideCompletionItems).
		 */
		range?: Range;

		/**
		 * @deprecated **Deprecated** in favor of `CompletionItem.insertText` and `CompletionItem.range`.
		 *
		 * ~~An [edit](#TextEdit) which is applied to a document when selecting
		 * this completion. When an edit is provided the value of
		 * [insertText](#CompletionItem.insertText) is ignored.~~
		 *
		 * ~~The [range](#Range) of the edit must be single-line and on the same
		 * line completions were [requested](#CompletionItemProvider.provideCompletionItems) at.~~
		 */
		textEdit?: TextEdit;

		/**
		 * An optional array of additional [text edits](#TextEdit) that are applied when
		 * selecting this completion. Edits must not overlap with the main [edit](#CompletionItem.textEdit)
		 * nor with themselves.
		 */
		additionalTextEdits?: TextEdit[];

		/**
		 * An optional [command](#Command) that is executed *after* inserting this completion. *Note* that
		 * additional modifications to the current document should be described with the
		 * [additionalTextEdits](#CompletionItem.additionalTextEdits)-property.
		 */
		command?: Command;

		/**
		 * Creates a new completion item.
		 *
		 * Completion items must have at least a [label](#CompletionItem.label) which then
		 * will be used as insert text as well as for sorting and filtering.
		 *
		 * @param label The label of the completion.
		 * @param kind The [kind](#CompletionItemKind) of the completion.
		 */
		constructor(label: string, kind?: CompletionItemKind);
	}
```

This is a breaking change in the LSP and the idea was to adopt this in LSP 3.x protocol version since it will force clients to honor the range attribute.

What is strange with the completion item in our case is that it only contains a range and not insertText. @aeschli any idea why?

**Comment by mniewrzal** [comment]
I checked completion result once again with vscode (trace option) and `insertText` is available:
```json
        {
            "label": "align-items",
            "documentation": "Aligns flex items along the cross axis of the current line of the flex container.\n(Edge, Firefox 22, Safari 9, Chrome 29, IE 11, Opera 12.1)",
            "insertText": "align-items",
            "range": {
                "start": {
                    "line": 1,
                    "character": 1
                },
                "end": {
                    "line": 1,
                    "character": 10
                }
            },
            "kind": 10
        },
```
Previously I pasted output from java LSP client, maybe because of changes in json structure it was not complete.

New CompletionItem structure looks better imho but it would be nice to honor both versions for some time :)

**Comment by dbaeumer** [comment]
IMO the CSS and HTML server need to raise the major version number to indicate that breakage or they need to send both [insertText, range] and textEdit.

@aeschli I assign this to you since there is nothing I from the LSP can do about this.

**Comment by mniewrzal** [comment]
Another example of breaking LS protocol is completion result for `background-color` (CSS):
```json
{
	"label": "hsla",
	"detail": "hsla($hue, $saturation, $lightness, $alpha)",
	"documentation": "Creates a Color from hue, saturation, lightness, and alpha values.",
	"insertText": {
		"value": "hsla(${1:hue}, ${2:saturation}, ${3:lightness}, ${4:alpha})"
	},
	"range": {
		"start": {
			"line": 1,
			"character": 18
		},
		"end": {
			"line": 1,
			"character": 18
		}
	},
	"kind": 3
},
```
Field `insertText` is a structure, **not** just string.

**Comment by aeschli** [comment]
The HTML, CSS and JSON language servers now use the upcoming 3.0.0-alpha.x version of the LSP.
As these servers are just part of our built-in extensions, the versions in the package.json's is not maintained. I actually don't think that a increase version would have helped much here:
@dbaeumer We should
- finalize and release 3.0.0
- add a version check as part of the protocol, e.g. on initialize. It should, IMO, be part of the LSP.

**Comment by dbaeumer** [comment]
I am in the process of finalizing this. When starting to write the spec I recognized some inconsistencies wich I started to clean out. 

An increased version would help if we would package our servers on npm. That way users would not get 'breaking' servers automatically.

Regarding version numbers: instead of version number I am using 'ClientCapabilities' feature flags now for this. I discussed this with the debugger people and they do the same and it is more flexible than protocol versions. The final 3.0 version will contain a flag: `ClientCapabilies.textDocument.completionItem.rangeProperty`. If false then a server should not use the new range property.

**Comment by aeschli** [comment]
Range property is only half of the change, the other one is the SnippetString for insertText.
This might lead to a lot of the flags and will make the server code complicated if I wanted to provide backward compatibility in various combinations. Note that my server uses the html-language-service for computing the proposals

**Comment by dbaeumer** [comment]
In the TS extension I use version numbers to make things backwards compatible which is even more complicated since I need to keep an 'inferred' list now which version has which features :-).

If you want to avoid complicated code you are free to check the capability flags in the initialize method and error the initialize call if you can't or want not to be backwards compatible. 

Overall I think this is more flexible and easier to handle than version numbers.

**Comment by dbaeumer** [comment]
I will add a flag for the SnippetString for insertText.

**Comment by dbaeumer** [comment]
I added a flag for this as well. To make this more flexible for the future I renamed the SnippetString to TypedString with a type property (Normal, Snippet). This will also address the problem of typed languages which don't want tot deal with or typing. They can always send a TypedString. 

Sorry for the breakage on our end but it should be very minimal.

**Comment by mniewrzal** [comment]
Yet another inconsistency I found in `ServerCapabilities` structure. LSP 3.0 spec describes `documentLinkProvider` as:
```
    /**
     * The server provides document link support.
     */
    documentLinkProvider?: DocumentLinkOptions;
```
but vscode HTML language server returns `boolean`:
```json
{
    "capabilities": {
        "textDocumentSync": 1,
        ...
        "hoverProvider": true,
        "documentHighlightProvider": true,
        "documentRangeFormattingProvider": true,
        "documentLinkProvider": true,
        "documentSymbolProvider": true,
       ...
    }
}
```

**Comment by dbaeumer** [comment]
@aeschli document links can have a resolve option. So `documentLinkProvider` is defined as 
```ts
export interface DocumentLinkOptions {
	/**
	 * Document links have a resolve provider as well.
	 */
	resolveProvider?: boolean;
}
```

**Comment by aeschli** [comment]
I fixed the documentLinkProvider. Not clear to me why typescript lets me use true here.

**Comment by aeschli** [comment]
The language servers now implement LSP 3.0. LSP 3.0 got fix to be backward compatible to LSP 1.0.
However, our language server implementation will not enable code completion and formating unless the client support snippets and supports dynamic capabilities.

**Comment by dbaeumer** [comment]
@aeschli thanks!
